### ISSUE TITLE
Segfault with regular expression

### ISSUE THREAD

**Issue by giggsey** [facebook/hhvm#5130]
Code:

```php
<?php

/*
require __DIR__ . '/vendor/autoload.php';
$phoneNumberUtil = \libphonenumber\PhoneNumberUtil::getInstance();
$number = $phoneNumberUtil->parse('011543549480042', 'US');
*/

$array = array();
$ret = preg_match('/^(?:          0?(?:            (11|             2(?:               2(?:                 02?|                 [13]|                 2[13-79]|                 4[1-6]|                 5[2457]|                 6[124-8]|                 7[1-4]|                 8[13-6]|                 9[1267]               )|               3(?:                 02?|                 1[467]|                 2[03-6]|                 3[13-8]|                 [49][2-6]|                 5[2-8]|                 [67]               )|               4(?:                 7[3-578]|                 9               )|               6(?:                 [0136]|                 2[24-6]|                 4[6-8]?|                 5[15-8]               )|               80|               9(?:                 0[1-3]|                 [19]|                 2\d|                 3[1-6]|                 4[02568]?|                 5[2-4]|                 6[2-46]|                 72?|                 8[23]?               )            )|            3(?:              3(?:                2[79]|                6|                8[2578]              )|              4(?:                0[124-9]|                [12]|                3[5-8]?|                4[24-7]|                5[4-68]?|                6[02-9]|                7[126]|                8[2379]?|                9[1-36-8]              )|              5(?:                1|                2[1245]|                3[237]?|                4[1-46-9]|                6[2-4]|                7[1-6]|                8[2-5]?              )|              6[24]|              7(?:                1[1568]|                2[15]|                3[145]|                4[13]|                5[14-8]|                [069]|                7[2-57]|                8[126]              )|              8(?:                [01]|                2[15-7]|                3[2578]?|                4[13-6]|                5[4-8]?|                6[1-357-9]|                7[36-8]?|                8[5-8]?|                9[124]              )            )          )?15        )?)/x', '3549480042', $array, PREG_OFFSET_CAPTURE);

var_dump($ret, $array);
```

When I run via `hhvm -m d i3.php`, I get:
```
Host: helms-deep
ProcessID: 15614
ThreadID: 7f3706ff1500
ThreadPID: 15614
Name: /usr/bin/hhvm
Type: Segmentation fault
Runtime: hhvm
Version: tags/HHVM-3.6.1-0-gc051feff88ba8050f33f10f7b1fb07f8d3415f9e
DebuggerCount: 1

Arguments: i3.php
ThreadType: CLI
IsDebugger: True

# 0  ?? at hhvm:0
```

With gdb:

```
(gdb) run i3.php
Starting program: /usr/bin/hhvm i3.php
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Program received signal SIGSEGV, Segmentation fault.
0x00007ffff7fe52ff in ?? ()
(gdb) #0  0x00007ffff7fe52ff in ?? ()
#1  0x00007ffff7fd6000 in ?? ()
#2  0x0000000000707a19 in ?? ()
#3  0x00007ffff5ce0980 in ?? () from /lib/x86_64-linux-gnu/libpcre.so.3
#4  0x0000000000402910 in ?? ()
#5  0x0000000100000000 in ?? ()
#6  0x00000000000f4240 in ?? ()
#7  0x00007ffff5ce0350 in ?? () from /lib/x86_64-linux-gnu/libpcre.so.3
#8  0x00007fffe000e238 in ?? ()
#9  0x00007fffe000e237 in ?? ()
#10 0x00007fffe000e238 in ?? ()
#11 0x00007fffe000e23c in ?? ()
#12 0x00007ffff7ffe520 in ?? ()
#13 0x00007ffff7ffe1c8 in _r_debug ()
#14 0x00007ffff7de4991 in _dl_lookup_symbol_x (undef_name=0x1453954 <HPHP::tvIsPlausible(HPHP::TypedValue)+233> "]\303UH\211\345@\200\376Pt\031\271`\217\310\002\272]", undef_map=0x7fffe000e242, ref=0x18, symbol_scope=
    0x240986d <HPHP::popRank(HPHP::Rank)+107>, version=0xf423e, type_class=-15392, flags=1000000, skip_map=0x5f00000027) at dl-lookup.c:737
#15 0x00007fffe7d09d70 in ?? ()
#16 0x00000000000f4240 in ?? ()
#17 0x0000005f00000027 in ?? ()
#18 0x0000006100000028 in ?? ()
#19 0x00000000000186a0 in ?? ()
#20 0x000000290000002e in ?? ()
#21 0x00007fffe0c4b7a0 in ?? ()
#22 0x00007fffffffc8d0 in ?? ()
#23 0xfcb885453cc85300 in ?? ()
#24 0x0000000000001753 in ?? ()
#25 0x00007fffe1400030 in ?? ()
#26 0x00007fffe1400000 in ?? ()
#27 0x0000000000000100 in ?? ()
#28 0x00007fffe541b080 in ?? ()
#29 0x00007fffe000e220 in ?? ()
#30 0x00007fffffffc8b0 in ?? ()
#31 0x000000000140df2d in HPHP::preg_match (pattern=..., subject=..., matches=..., flags=<optimised out>, offset=<optimised out>) at /tmp/tmp.p6bqC6bj1v/hphp/runtime/base/preg.cpp:1119
Backtrace stopped: previous frame inner to this frame (corrupt stack?)

```

When running on 3v4l, it segfaults in some PHP7 versions (not in March+April's release), and all of the HHVM versions they have: http://3v4l.org/IlQrt

(Part of giggsey/libphonenumber-for-php/issues/54)

**Comment by paulbiss** [comment]
It looks a lot like the crash is in the pcre library.

**Comment by dpaneda** [comment]
Same problem here with a similar code, on a jessie machine.

```
<?php

$pattern = '0?(?:(11|2(?:2(?:02?|[13]|2[13-79]|4[1-6]|5[2457]|6[124-8]|7[1-4]|8[13-6]|9[1267])|3(?:02?|1[467]|2[03-6]|3[13-8]|[49][2-6]|5[2-8]|[67])|4(?:7[3-578]|9)|6(?:[0136]|2[24-6]|4[6-8]?|5[15-8])|80|9(?:0[1-3]|[19]|2\d|3[1-6]|4[02568]?|5[2-4]|6[2-46]|72?|8[23]?))|3(?:3(?:2[79]|6|8[2578])|4(?:0[124-9]|[12]|3[5-8]?|4[24-7]|5[4-68]?|6[02-9]|7[126]|8[2379]?|9[1-36-8])|5(?:1|2[1245]|3[237]?|4[1-46-9]|6[2-4]|7[1-6]|8[2-5]?)|6[24]|7(?:1[1568]|2[15]|3[145]|4[13]|5[14-8]|[069]|7[2-57]|8[126])|8(?:[01]|2[15-7]|3[2578]?|4[13-6]|5[4-8]?|6[1-357-9]|7[36-8]?|8[5-8]?|9[124])))15)?';

$a = preg_match("/$pattern/", '2966');
var_dump($a);
```

It looks like a problem on the pcre, because it's failing with 8.35 but fixed on 8.37, but I'm not able to reproduce the problem using exactly the same parameters calling pcre directly from a C program, so I'm not really sure.

**Comment by Orvid** [comment]
It doesn't segfault on any current versions of HHVM.
